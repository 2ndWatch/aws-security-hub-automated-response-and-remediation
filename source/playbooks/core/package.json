{
  "name": "playbook-construct",
  "version": "1.0.1",
  "bin": {
    "playbook-construct": "bin/playbook-constuct.js"
  },
  "scripts": {
    "build": "tsc",
    "watch": "tsc -w",
    "test": "jest",
    "cdk": "cdk"
  },
  "devDependencies": {
    "@aws-cdk/assert": "1.60.0",
    "@types/jest": "^25.2.1",
    "@types/node": "13.13.4",
    "jest": "^25.5.4",
    "ts-jest": "^25.3.1",
    "aws-cdk": "1.60.0",
    "ts-node": "^8.1.0",
    "typescript": "~3.7.2"
  },
  "dependencies": {
    "@aws-cdk/aws-events": "1.60.0",
    "@aws-cdk/aws-events-targets": "1.60.0",
    "@aws-cdk/aws-iam": "1.60.0",
    "@aws-cdk/aws-kms": "1.60.0",
    "@aws-cdk/aws-lambda": "1.60.0",
    "@aws-cdk/aws-logs": "1.60.0",
    "@aws-cdk/aws-s3": "1.60.0",
    "@aws-cdk/aws-securityhub": "1.60.0",
    "@aws-cdk/aws-servicecatalog": "1.60.0",
    "@aws-cdk/aws-sns": "1.60.0",
    "@aws-cdk/aws-ssm": "1.60.0",
    "@aws-cdk/core": "1.60.0",
    "@aws-cdk/custom-resources": "1.60.0",
    "@types/js-yaml": "^3.12.3",
    "source-map-support": "^0.5.16",
    "@aws-solutions-constructs/aws-events-rule-lambda": "1.60.0"
  }
}
