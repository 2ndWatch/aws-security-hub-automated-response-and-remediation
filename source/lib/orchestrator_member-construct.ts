#!/usr/bin/env node
/*****************************************************************************
 *  Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.   *
 *                                                                            *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may   *
 *  not use this file except in compliance with the License. A copy of the    *
 *  License is located at                                                     *
 *                                                                            *
 *      http://www.apache.org/licenses/LICENSE-2.0                            *
 *                                                                            *
 *  or in the 'license' file accompanying this file. This file is distributed *
 *  on an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,        *
 *  express or implied. See the License for the specific language governing   *
 *  permissions and limitations under the License.                            *
 *****************************************************************************/

// Create a new stack to deploy a Playbook in a member account
// Orchestrator member role
// Remediation role(s)
// SSM Automation Documents

import * as cdk from '@aws-cdk/core';
import { 
    PolicyStatement, 
    Effect, 
    Role, 
    PolicyDocument, 
    ArnPrincipal, 
    CompositePrincipal, 
    CfnRole 
} from '@aws-cdk/aws-iam';

export interface SHARRStackProps extends cdk.StackProps  {
    solutionId: string;
    solutionVersion: string;
    solutionDistBucket: string;
    solutionDistName: string;
    solutionName: string;
    securityStandard: string;
}

export class MemberStack extends cdk.Stack {

  SEND_ANONYMOUS_DATA = 'Yes'
  public readonly adminAccountNumber: cdk.CfnParameter;

  constructor(scope: cdk.App, id: string, props: SHARRStackProps) {
    super(scope, id, props);

    const validAwsAccount = "\\d{12}"
    const adminRoleName = props.solutionId + '-SHARR-Orchestrator-Admin_' + this.region

    this.adminAccountNumber = new cdk.CfnParameter(this, 'Admin Account Number', {
        description: "Admin account number",
        type: "String",
        allowedPattern: validAwsAccount
    });

    const memberPolicy = new PolicyDocument();
    /**
     * @description Cross-account permissions for Orchestration role
     * @type {PolicyStatement}
     */
    const ssmROPerms = new PolicyStatement()
    ssmROPerms.addActions("ssm:DescribeAutomationExecutions")
    ssmROPerms.addActions("ssm:DescribeDocument")
    ssmROPerms.effect = Effect.ALLOW;
    ssmROPerms.addResources(
        "arn:" + this.partition + ":ssm:" + this.region + ":*:*"
    )
    memberPolicy.addStatements(ssmROPerms)

    let principalPolicyStatement = new PolicyStatement();

    principalPolicyStatement.addActions("sts:AssumeRole");
    principalPolicyStatement.effect = Effect.ALLOW;

    let roleprincipal = new ArnPrincipal(
        'arn:' + this.partition + ':iam::' + this.adminAccountNumber.value + ':role/' +
        adminRoleName
    );

    let principals = new CompositePrincipal(roleprincipal);
    principals.addToPolicy(principalPolicyStatement);

    let memberRole = new Role(this, 'MemberAccountRole', {
        assumedBy: principals,
        inlinePolicies: {
            'default_lambdaPolicy': memberPolicy
        },
        roleName: props.solutionId + '-SHARR-Orchestrator-Member-' + props.securityStandard + '_' + this.region
    });

    const memberRoleResource = memberRole.node.findChild('Resource') as CfnRole;

    memberRoleResource.cfnOptions.metadata = {
        cfn_nag: {
            rules_to_suppress: [{
                id: 'W11',
                reason: 'Resource * is required due to the administrative nature of the solution.'
            },{
                id: 'W28',
                reason: 'Static names chosen intentionally to provide integration in cross-account permissions'
            }]
        }
    };
  }
}


