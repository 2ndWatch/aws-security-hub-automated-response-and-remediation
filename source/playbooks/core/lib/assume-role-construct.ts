/*****************************************************************************
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.   *
 *                                                                            *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may   *
 *  not use this file except in compliance with the License. A copy of the    *
 *  License is located at                                                     *
 *                                                                            *
 *      http://www.apache.org/licenses/LICENSE-2.0                            *
 *                                                                            *
 *  or in the 'license' file accompanying this file. This file is distributed *
 *  on an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,        *
 *  express or implied. See the License for the specific language governing   *
 *  permissions and limitations under the License.                            *
 *****************************************************************************/

import * as cdk from '@aws-cdk/core';
import { PolicyStatement, Effect, Role, PolicyDocument, ArnPrincipal, ServicePrincipal, CompositePrincipal, CfnRole } from '@aws-cdk/aws-iam';

export interface AssumeRoleConstructProps {
    masterAccountNumber: cdk.CfnParameter,
    solutionId: string,
    lambdaPolicy: PolicyDocument,
    lambdaHandlerName: string
    service?: string
}

export class AssumeRoleConstruct extends cdk.Construct {

    constructor(scope: cdk.Construct, id: string, props: AssumeRoleConstructProps) {
        super(scope, id);

        //DEFINE PRINCIPLE RESOURCE WHICH WILL ASSUME THE ROLE.

        let principalPolicyStatement = new PolicyStatement();
        principalPolicyStatement.addActions("sts:AssumeRole");
        principalPolicyStatement.effect = Effect.ALLOW;

        let roleprincipal = new ArnPrincipal('arn:aws:iam::' + props.masterAccountNumber.value + ':role/' +
            props.solutionId + '_' + props.lambdaHandlerName +  '_lambdaRole');

        let principals = new CompositePrincipal(roleprincipal);
        principals.addToPolicy(principalPolicyStatement);

        if (props.service) {
            let serviceprincipal = new ServicePrincipal(props.service)
            principals.addPrincipals(serviceprincipal);
        }

        let lambdaRole = new Role(this, 'MemberAccountRole', {
            assumedBy: principals,
            inlinePolicies: {
                'default_lambdaPolicy': props.lambdaPolicy
            },
            roleName: props.solutionId + '_' + props.lambdaHandlerName + '_memberRole'
        });

        const lambdaRoleResource = lambdaRole.node.findChild('Resource') as CfnRole;

        lambdaRoleResource.cfnOptions.metadata = {
            cfn_nag: {
                rules_to_suppress: [{
                    id: 'W11',
                    reason: 'Resource * is required due to the administrative nature of the solution.'
                },{
                    id: 'W28',
                    reason: 'Static names chosen intentionally to provide integration in cross-account permissions'
                }]
            }
        };
    }
}