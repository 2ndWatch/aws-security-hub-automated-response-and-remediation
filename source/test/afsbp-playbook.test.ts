#!/usr/bin/env node
/*****************************************************************************
 *  Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.   *
 *                                                                            *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may   *
 *  not use this file except in compliance with the License. A copy of the    *
 *  License is located at                                                     *
 *                                                                            *
 *      http://www.apache.org/licenses/LICENSE-2.0                            *
 *                                                                            *
 *  or in the 'license' file accompanying this file. This file is distributed *
 *  on an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,        *
 *  express or implied. See the License for the specific language governing   *
 *  permissions and limitations under the License.                            *
 *****************************************************************************/

import { SynthUtils } from '@aws-cdk/assert';
import '@aws-cdk/assert/jest';
import { BlockPublicAccess, Bucket, BucketAccessControl, BucketEncryption } from '@aws-cdk/aws-s3';
import { App, Stack } from '@aws-cdk/core';
import { OrchestratorConstruct } from '../playbooks/AFSBP/lib/afsbp-orchestrator-construct';

test('test App AFSBP Orchestrator Construct', () => {

    const app = new App();
    const stack = new Stack(app, 'testStack', {
        stackName: 'testStack'
    });

    new OrchestratorConstruct(stack, 'Orchestrator', {
	    roleArn: 'arn:aws-test:iam::111111111111:role/TestRole',
	    ssmDocStateLambda: 'xxx',
	    ssmExecDocLambda: 'yyy',
	    ssmExecMonitorLambda: 'zzz',
	    notifyLambda: 'aaa',
        solutionId: 'bbb'
    });

    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});