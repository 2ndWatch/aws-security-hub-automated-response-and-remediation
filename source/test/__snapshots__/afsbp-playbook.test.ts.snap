// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test App AFSBP Orchestrator Construct 1`] = `
Object {
  "Conditions": Object {
    "OrchestratorisGovCloudDE3D4910": Object {
      "Fn::Not": Array [
        Object {
          "Condition": "OrchestratorisNotGovCloud793D69A3",
        },
      ],
    },
    "OrchestratorisNotGovCloud793D69A3": Object {
      "Fn::Not": Array [
        Object {
          "Fn::Equals": Array [
            Object {
              "Ref": "AWS::Partition",
            },
            "aws-us-gov",
          ],
        },
      ],
    },
  },
  "Resources": Object {
    "OrchestratorRole9CF251DB": Object {
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W11",
              "reason": "CloudWatch Logs permissions require resource * except for DescribeLogGroups, except for GovCloud, which only works with resource *",
            },
          ],
        },
      },
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "states.",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ".amazonaws.com",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "logs:CreateLogDelivery",
                    "logs:GetLogDelivery",
                    "logs:UpdateLogDelivery",
                    "logs:DeleteLogDelivery",
                    "logs:ListLogDeliveries",
                    "logs:PutResourcePolicy",
                    "logs:DescribeResourcePolicies",
                    "logs:DescribeLogGroups",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                },
                Object {
                  "Action": "lambda:InvokeFunction",
                  "Effect": "Allow",
                  "Resource": Array [
                    Object {
                      "Fn::Join": Array [
                        "",
                        Array [
                          "arn:",
                          Object {
                            "Ref": "AWS::Partition",
                          },
                          ":lambda:",
                          Object {
                            "Ref": "AWS::Region",
                          },
                          ":",
                          Object {
                            "Ref": "AWS::AccountId",
                          },
                          ":function:undefined",
                        ],
                      ],
                    },
                    Object {
                      "Fn::Join": Array [
                        "",
                        Array [
                          "arn:",
                          Object {
                            "Ref": "AWS::Partition",
                          },
                          ":lambda:",
                          Object {
                            "Ref": "AWS::Region",
                          },
                          ":",
                          Object {
                            "Ref": "AWS::AccountId",
                          },
                          ":function:undefined",
                        ],
                      ],
                    },
                    Object {
                      "Fn::Join": Array [
                        "",
                        Array [
                          "arn:",
                          Object {
                            "Ref": "AWS::Partition",
                          },
                          ":lambda:",
                          Object {
                            "Ref": "AWS::Region",
                          },
                          ":",
                          Object {
                            "Ref": "AWS::AccountId",
                          },
                          ":function:undefined",
                        ],
                      ],
                    },
                    Object {
                      "Fn::Join": Array [
                        "",
                        Array [
                          "arn:",
                          Object {
                            "Ref": "AWS::Partition",
                          },
                          ":lambda:",
                          Object {
                            "Ref": "AWS::Region",
                          },
                          ":",
                          Object {
                            "Ref": "AWS::AccountId",
                          },
                          ":function:undefined",
                        ],
                      ],
                    },
                  ],
                },
                Object {
                  "Action": Array [
                    "kms:Encrypt",
                    "kms:Decrypt",
                    "kms:GenerateDataKey",
                  ],
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "arn:",
                        Object {
                          "Ref": "AWS::Partition",
                        },
                        ":kms:",
                        Object {
                          "Ref": "AWS::Region",
                        },
                        ":",
                        Object {
                          "Ref": "AWS::AccountId",
                        },
                        ":alias/SO0111-SHARR-Key",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "BasePolicy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "OrchestratorSO0111SHARROrchestratorAFSBP18511A69": Object {
      "Condition": "OrchestratorisGovCloudDE3D4910",
      "DeletionPolicy": "Retain",
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W84",
              "reason": "KmsKeyId is not supported in GovCloud.",
            },
          ],
        },
      },
      "Properties": Object {
        "LogGroupName": "SO0111-SHARR-Orchestrator-AFSBP",
        "RetentionInDays": 365,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "OrchestratorSO0111SHARROrchestratorAFSBPEncrypted7E46D101": Object {
      "Condition": "OrchestratorisNotGovCloud793D69A3",
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "KmsKeyId": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":kms:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":alias/SO0111-SHARR-Key",
            ],
          ],
        },
        "LogGroupName": "SO0111-SHARR-Orchestrator-AFSBP",
        "RetentionInDays": 365,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "OrchestratorStateMachine1E795392": Object {
      "DependsOn": Array [
        "OrchestratorRole9CF251DB",
      ],
      "Properties": Object {
        "DefinitionString": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"StartAt\\":\\"Get Finding Data from Input\\",\\"States\\":{\\"Get Finding Data from Input\\":{\\"Type\\":\\"Pass\\",\\"Comment\\":\\"Extract top-level data needed for remediation\\",\\"Parameters\\":{\\"EventType.$\\":\\"$.detail-type\\",\\"Findings.$\\":\\"$.detail.findings\\",\\"Metrics\\":{\\"sendAnonymousMetrics.$\\":\\"States.Format('{{resolve:ssm:/Solutions/SO0111/sendAnonymousMetrics:1}}')\\"}},\\"Next\\":\\"Process Findings\\"},\\"Process Findings\\":{\\"Type\\":\\"Map\\",\\"Comment\\":\\"Process all findings in CloudWatch Event\\",\\"Next\\":\\"EOJ\\",\\"Parameters\\":{\\"Metrics.$\\":\\"$.Metrics\\",\\"Finding.$\\":\\"$$.Map.Item.Value\\",\\"EventType.$\\":\\"$.EventType\\"},\\"Iterator\\":{\\"StartAt\\":\\"Finding Workflow State NEW?\\",\\"States\\":{\\"Finding Workflow State NEW?\\":{\\"Type\\":\\"Choice\\",\\"Choices\\":[{\\"Or\\":[{\\"Variable\\":\\"$.EventType\\",\\"StringEquals\\":\\"Security Hub Findings - Custom Action\\"},{\\"And\\":[{\\"Variable\\":\\"$.Finding.Workflow.Status\\",\\"StringEquals\\":\\"NEW\\"},{\\"Variable\\":\\"$.EventType\\",\\"StringEquals\\":\\"Security Hub Findings - Imported\\"}]}],\\"Next\\":\\"Correct Security Standard?\\"}],\\"Default\\":\\"Finding Workflow State is not NEW\\"},\\"Finding Workflow State is not NEW\\":{\\"Type\\":\\"Pass\\",\\"Parameters\\":{\\"Notification\\":{\\"Message.$\\":\\"States.Format('Finding Workflow State is not NEW ({}).', $.Finding.Workflow.Status)\\",\\"State.$\\":\\"States.Format('NOTNEW')\\"},\\"Metrics.$\\":\\"$.Metrics\\",\\"EventType.$\\":\\"$.EventType\\",\\"Finding.$\\":\\"$.Finding\\"},\\"Next\\":\\"notify\\"},\\"notify\\":{\\"End\\":true,\\"Retry\\":[{\\"ErrorEquals\\":[\\"Lambda.ServiceException\\",\\"Lambda.AWSLambdaException\\",\\"Lambda.SdkClientException\\"],\\"IntervalSeconds\\":2,\\"MaxAttempts\\":6,\\"BackoffRate\\":2}],\\"Type\\":\\"Task\\",\\"Comment\\":\\"Send notifications\\",\\"TimeoutSeconds\\":300,\\"HeartbeatSeconds\\":60,\\"Resource\\":\\"arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":states:::lambda:invoke\\",\\"Parameters\\":{\\"FunctionName\\":\\"aaa\\",\\"Payload.$\\":\\"$\\"}},\\"Wrong Security Standard\\":{\\"Type\\":\\"Pass\\",\\"Parameters\\":{\\"Notification\\":{\\"Message.$\\":\\"States.Format('Finding Security Standard is not AFSBP ({}). Verify that the correct remediation was selected.', $.Finding.GeneratorId)\\",\\"State.$\\":\\"States.Format('WRONGSTANDARD')\\"},\\"Metrics.$\\":\\"$.Metrics\\",\\"EventType.$\\":\\"$.EventType\\",\\"Finding.$\\":\\"$.Finding\\"},\\"Next\\":\\"notify\\"},\\"Automation Document is not Active\\":{\\"Type\\":\\"Pass\\",\\"Parameters\\":{\\"Notification\\":{\\"Message.$\\":\\"States.Format('Automation Document ({}) is not active ({}) in the member account({}).', $.AutomationDocId, $.AutomationDocument.DocState, $.Finding.AwsAccountId)\\",\\"State.$\\":\\"States.Format('REMEDIATIONNOTACTIVE')\\",\\"updateSecHub\\":\\"yes\\"},\\"Metrics.$\\":\\"$.Metrics\\",\\"EventType.$\\":\\"$.EventType\\",\\"Finding.$\\":\\"$.Finding\\",\\"AccountId.$\\":\\"$.AccountId\\",\\"AutomationDocId.$\\":\\"$.AutomationDocId\\",\\"RemediationRole.$\\":\\"$.RemediationRole\\",\\"ControlId.$\\":\\"$.ControlId\\",\\"SecurityStandard.$\\":\\"$.SecurityStandard\\"},\\"Next\\":\\"notify\\"},\\"Orchestrator Failed\\":{\\"Type\\":\\"Pass\\",\\"Parameters\\":{\\"Notification\\":{\\"Message.$\\":\\"States.Format('Orchestrator failed: {}', $.Error)\\",\\"State.$\\":\\"States.Format('LAMBDAERROR')\\",\\"Details.$\\":\\"States.Format('Cause: {}', $.Cause)\\"},\\"SecurityStandard.$\\":\\"States.Format('AFSBP')\\",\\"Payload.$\\":\\"$\\"},\\"Next\\":\\"notify\\"},\\"Get Automation Document State\\":{\\"Next\\":\\"Automation Doc Active?\\",\\"Retry\\":[{\\"ErrorEquals\\":[\\"Lambda.ServiceException\\",\\"Lambda.AWSLambdaException\\",\\"Lambda.SdkClientException\\"],\\"IntervalSeconds\\":2,\\"MaxAttempts\\":6,\\"BackoffRate\\":2}],\\"Catch\\":[{\\"ErrorEquals\\":[\\"States.ALL\\"],\\"Next\\":\\"Orchestrator Failed\\"}],\\"Type\\":\\"Task\\",\\"Comment\\":\\"Get the status of the remediation automation document in the target account\\",\\"TimeoutSeconds\\":60,\\"ResultPath\\":\\"$.AutomationDocument\\",\\"Resource\\":\\"arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":states:::lambda:invoke\\",\\"Parameters\\":{\\"FunctionName\\":\\"xxx\\",\\"Payload.$\\":\\"$\\"},\\"ResultSelector\\":{\\"DocState.$\\":\\"$.Payload.status\\",\\"Message.$\\":\\"$.Payload.message\\"}},\\"extractParameters\\":{\\"Type\\":\\"Pass\\",\\"Comment\\":\\"Extract parameters needed for remediation\\",\\"Parameters\\":{\\"SecurityStandard.$\\":\\"States.Format('AFSBP')\\",\\"ControlId.$\\":\\"$.Finding.ProductFields.ControlId\\",\\"AccountId.$\\":\\"$.Finding.AwsAccountId\\",\\"Finding.$\\":\\"$.Finding\\",\\"RemediationRole.$\\":\\"States.Format('SO0111-SHARR-Remediation-AFSBP-{}', $.Finding.ProductFields.ControlId)\\",\\"AutomationDocId.$\\":\\"States.Format('SHARR_Remediation_AFSBP_{}', $.Finding.ProductFields.ControlId)\\",\\"Metrics.$\\":\\"$.Metrics\\",\\"EventType.$\\":\\"$.EventType\\"},\\"Next\\":\\"Get Automation Document State\\"},\\"Correct Security Standard?\\":{\\"Type\\":\\"Choice\\",\\"Choices\\":[{\\"Variable\\":\\"$.Finding.GeneratorId\\",\\"StringMatches\\":\\"aws-foundational-security-best-practices/v/1.0.0/*\\",\\"Next\\":\\"extractParameters\\"}],\\"Default\\":\\"Wrong Security Standard\\"},\\"Automation Doc Active?\\":{\\"Type\\":\\"Choice\\",\\"Choices\\":[{\\"Variable\\":\\"$.AutomationDocument.DocState\\",\\"StringEquals\\":\\"ACTIVE\\",\\"Next\\":\\"Execute Remediation\\"}],\\"Default\\":\\"Automation Document is not Active\\"},\\"Execute Remediation\\":{\\"Next\\":\\"execMonitor\\",\\"Retry\\":[{\\"ErrorEquals\\":[\\"Lambda.ServiceException\\",\\"Lambda.AWSLambdaException\\",\\"Lambda.SdkClientException\\"],\\"IntervalSeconds\\":2,\\"MaxAttempts\\":6,\\"BackoffRate\\":2}],\\"Catch\\":[{\\"ErrorEquals\\":[\\"States.ALL\\"],\\"Next\\":\\"Orchestrator Failed\\"}],\\"Type\\":\\"Task\\",\\"Comment\\":\\"Execute the SSM Automation Document in the target account\\",\\"TimeoutSeconds\\":300,\\"HeartbeatSeconds\\":60,\\"ResultPath\\":\\"$.SSMExecution\\",\\"Resource\\":\\"arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":states:::lambda:invoke\\",\\"Parameters\\":{\\"FunctionName\\":\\"yyy\\",\\"Payload.$\\":\\"$\\"},\\"ResultSelector\\":{\\"ExecState.$\\":\\"$.Payload.status\\",\\"Message.$\\":\\"$.Payload.message\\",\\"ExecId.$\\":\\"$.Payload.executionid\\"}},\\"execMonitor\\":{\\"Next\\":\\"Remediation completed?\\",\\"Retry\\":[{\\"ErrorEquals\\":[\\"Lambda.ServiceException\\",\\"Lambda.AWSLambdaException\\",\\"Lambda.SdkClientException\\"],\\"IntervalSeconds\\":2,\\"MaxAttempts\\":6,\\"BackoffRate\\":2}],\\"Catch\\":[{\\"ErrorEquals\\":[\\"States.ALL\\"],\\"Next\\":\\"Orchestrator Failed\\"}],\\"Type\\":\\"Task\\",\\"Comment\\":\\"Monitor the remediation execution until done\\",\\"TimeoutSeconds\\":300,\\"HeartbeatSeconds\\":60,\\"ResultPath\\":\\"$.Remediation\\",\\"Resource\\":\\"arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":states:::lambda:invoke\\",\\"Parameters\\":{\\"FunctionName\\":\\"zzz\\",\\"Payload.$\\":\\"$\\"},\\"ResultSelector\\":{\\"ExecState.$\\":\\"$.Payload.status\\",\\"ExecId.$\\":\\"$.Payload.executionid\\",\\"RemediationState.$\\":\\"$.Payload.remediation_status\\",\\"Message.$\\":\\"$.Payload.message\\",\\"LogData.$\\":\\"$.Payload.logdata\\",\\"AffectedObject.$\\":\\"$.Payload.affected_object\\"}},\\"Wait for Remediation\\":{\\"Type\\":\\"Wait\\",\\"Seconds\\":15,\\"Next\\":\\"execMonitor\\"},\\"Remediation completed?\\":{\\"Type\\":\\"Choice\\",\\"Choices\\":[{\\"Variable\\":\\"$.Remediation.RemediationState\\",\\"StringEquals\\":\\"Failed\\",\\"Next\\":\\"Remediation Failed\\"},{\\"Variable\\":\\"$.Remediation.ExecState\\",\\"StringEquals\\":\\"Success\\",\\"Next\\":\\"Remediation Succeeded\\"},{\\"Variable\\":\\"$.Remediation.ExecState\\",\\"StringEquals\\":\\"TimedOut\\",\\"Next\\":\\"Remediation Failed\\"},{\\"Variable\\":\\"$.Remediation.ExecState\\",\\"StringEquals\\":\\"Cancelling\\",\\"Next\\":\\"Remediation Failed\\"},{\\"Variable\\":\\"$.Remediation.ExecState\\",\\"StringEquals\\":\\"Cancelled\\",\\"Next\\":\\"Remediation Failed\\"},{\\"Variable\\":\\"$.Remediation.ExecState\\",\\"StringEquals\\":\\"Failed\\",\\"Next\\":\\"Remediation Failed\\"}],\\"Default\\":\\"Wait for Remediation\\"},\\"Remediation Failed\\":{\\"Type\\":\\"Pass\\",\\"Comment\\":\\"Set parameters for notification\\",\\"Parameters\\":{\\"Metrics.$\\":\\"$.Metrics\\",\\"EventType.$\\":\\"$.EventType\\",\\"Finding.$\\":\\"$.Finding\\",\\"AccountId.$\\":\\"$.AccountId\\",\\"AutomationDocId.$\\":\\"$.AutomationDocId\\",\\"RemediationRole.$\\":\\"$.RemediationRole\\",\\"ControlId.$\\":\\"$.ControlId\\",\\"SecurityStandard.$\\":\\"$.SecurityStandard\\",\\"Notification\\":{\\"Message.$\\":\\"States.Format('Remediation failed for {} control {} in account {}: {}', $.SecurityStandard, $.ControlId, $.AccountId, $.Remediation.Message)\\",\\"State.$\\":\\"$.Remediation.ExecState\\",\\"Details.$\\":\\"$.Remediation.LogData\\",\\"ExecId.$\\":\\"$.Remediation.ExecId\\",\\"AffectedObject.$\\":\\"$.Remediation.AffectedObject\\"}},\\"Next\\":\\"notify\\"},\\"Remediation Succeeded\\":{\\"Type\\":\\"Pass\\",\\"Comment\\":\\"Set parameters for notification\\",\\"Parameters\\":{\\"Metrics.$\\":\\"$.Metrics\\",\\"EventType.$\\":\\"$.EventType\\",\\"Finding.$\\":\\"$.Finding\\",\\"AccountId.$\\":\\"$.AccountId\\",\\"AutomationDocId.$\\":\\"$.AutomationDocId\\",\\"RemediationRole.$\\":\\"$.RemediationRole\\",\\"ControlId.$\\":\\"$.ControlId\\",\\"SecurityStandard.$\\":\\"$.SecurityStandard\\",\\"Notification\\":{\\"Message.$\\":\\"States.Format('Remediation succeeded for {} control {} in account {}: {}', $.SecurityStandard, $.ControlId, $.AccountId, $.Remediation.Message)\\",\\"State.$\\":\\"States.Format('SUCCESS')\\",\\"Details.$\\":\\"$.Remediation.LogData\\",\\"ExecId.$\\":\\"$.Remediation.ExecId\\",\\"AffectedObject.$\\":\\"$.Remediation.AffectedObject\\"}},\\"Next\\":\\"notify\\"}}},\\"ItemsPath\\":\\"$.Findings\\"},\\"EOJ\\":{\\"Type\\":\\"Pass\\",\\"Comment\\":\\"END-OF-JOB\\",\\"End\\":true}},\\"TimeoutSeconds\\":900}",
            ],
          ],
        },
        "LoggingConfiguration": Object {
          "Destinations": Array [
            Object {
              "CloudWatchLogsLogGroup": Object {
                "LogGroupArn": Object {
                  "Fn::If": Array [
                    "OrchestratorisNotGovCloud793D69A3",
                    Object {
                      "Fn::GetAtt": Array [
                        "OrchestratorSO0111SHARROrchestratorAFSBPEncrypted7E46D101",
                        "Arn",
                      ],
                    },
                    Object {
                      "Fn::GetAtt": Array [
                        "OrchestratorSO0111SHARROrchestratorAFSBP18511A69",
                        "Arn",
                      ],
                    },
                  ],
                },
              },
            },
          ],
          "IncludeExecutionData": true,
          "Level": "ALL",
        },
        "RoleArn": Object {
          "Fn::GetAtt": Array [
            "OrchestratorRole9CF251DB",
            "Arn",
          ],
        },
        "StateMachineName": "SO0111-SHARR-Orchestrator-AFSBP",
      },
      "Type": "AWS::StepFunctions::StateMachine",
    },
  },
}
`;
